### PR TITLE
What's an Outbox and why do we need it?

### PR DESCRIPTION
**Pull request by mauroservienti**
## PoA

- [x] Synopsis
- [x] Tags
- [x] Header image
- [x] Header image credits
- [x] Review
- [x] Rename file to match publish date
- [ ] Buffer content

### CODE REVIEW

File: _posts/outbox-what-and-why.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+It comes at a cost, though. Transactions are expensive, and distributed transactions are even more costly. They have a performance impact because getting all the participants into an agreement takes time. As you can imagine, the more participants, the greater the effect. Last but not least, transaction protocols rely on high network stability. Poor network connectivity hinders participants' ability to participate in the distributed transaction coordination process. That means it's unlikely cloud environments support distributed transaction protocols.
+
+All that means the above-shown diagram only works in a cloud environment if the database server is local to the same machine where the MSMQ is. This sounds illogical because that means the only option to use distributed transactions is to have all participants on the same box, which is precisely what we don't want.
```
Comment (kbaley) [inline _posts/outbox-what-and-why.md:27]:
```suggestion
All this means that the above diagram only works in a cloud environment if the database server is local to the same machine where the MSMQ machine is. This sounds illogical because that means the only option to use distributed transactions is to have all participants on the same box, which is precisely what we don't want in the cloud.
```

File: _posts/outbox-what-and-why.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+Let's dissect that. We could be in one of the following scenarios:
+
+<div class="mermaid">
+graph LR
+    E[Endpoint] -- store data --> DB[(database)]
+    E -- send messages --> Q[Queuing system]
+</div>
+
+In this first one, there isn't an incoming message. For example, an HTTP request could trigger the endpoint to send out the message.
```
Comment (kbaley) [inline _posts/outbox-what-and-why.md:41]:
```suggestion
In the first one, there isn't an incoming message. For example, an HTTP request could trigger the endpoint to send out the message.
```

File: _posts/outbox-what-and-why.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+We could be doing the same thing for the "regular" inbox/outbox scenario with an incoming message with a few tweaks. Let's recap the outbox steps first:
+
+1. Start a local transaction and store the message id in the outbox storage.
+2. Handle the message from the business logic perspective and store any message dispatch intent.
+3. Commit the local transaction.
+4. Send messages we want to send and mark them as sent (outside the original local transaction).
+5. Acknowledge the incoming message, deleting it from the queue.
+
+Not having an incoming message changes things a bit. We need a way to start a local transaction to store data and dispatch intentions. If we cannot use the incoming message to trigger the local transaction, we could use an ASP.Net middleware that begins the local transaction when the HTTP request starts. We can use the same middleware to decide when to commit the transaction, and that is when the HTTP request finishes. With that set, we can tweak the endpoint steps in the following way:
```
Comment (kbaley) [inline _posts/outbox-what-and-why.md:121]:
```suggestion
Not having an incoming message changes things a bit. We need a way to start a local transaction to store data and dispatch intentions. If we cannot use the incoming message to trigger the local transaction, we could use an ASP.NET middleware that begins the local transaction when the HTTP request starts. We can use the same middleware to decide when to commit the transaction, and that is when the HTTP request finishes. With that set, we can tweak the endpoint steps in the following way:
```

File: _posts/outbox-what-and-why.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+Now, let's look again at failure scenarios for other steps of the process:
+
+- Step one, two, or three fail: We do nothing; the HTTP request's endpoint allows the exception to bubble up to the HTTP client, for example, with an HTTP 500. It's a common synchronous request/response failing scenario. Since it happens inside a transaction, the endpoint rolls back every change, and the system remains consistent and will dispatch no messages.
+- Step four fails: Similarly to the previous scenario, if the endpoint fails in committing the local transaction, everything is left unchanged, and an error will bubble up to the client. However, the control message will arrive at the local queue. It'll find nothing because there were no changes. It'll retry a few times and eventually end up in the error queue, where it can be safely discarded.
```
Comment (kbaley) [inline _posts/outbox-what-and-why.md:143]:
```suggestion
- Step four fails: Similar to the previous scenario, if the endpoint fails in committing the local transaction, everything is left unchanged, and an error will bubble up to the client. However, the control message will arrive at the local queue. It'll find nothing because there were no changes. It'll retry a few times and eventually end up in the error queue, where it can be safely discarded.
```

File: _posts/outbox-what-and-why.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+- Step five fails: The control message will retry dispatching outgoing messages. As said, it's a regular outbox at this point.
```
Comment (kbaley) [inline _posts/outbox-what-and-why.md:144]:
```suggestion
- Step five fails: The control message will retry dispatching outgoing messages. As we saw earlier, it's a regular outbox at this point.
```

File: _posts/outbox-what-and-why.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+## Conclusion
+
+When using a queueing system and at the same time storing data, we need to guarantee consistency. We don't want to store data and fail in dispatching messages, nor do we wish to deliver messages and fail in storing data. We relied on distributed transactions for a long time to achieve the desired goal. However, most queuing infrastructures and cloud-enabled storages don't support distributed transactions. It takes a lot of work to implement, especially if there is a need to support multiple storage options and queuing systems. That's one of the reasons to rely upon a dedicated toolkit. For more information, please refer to the following articles:
```
Comment (kbaley) [inline _posts/outbox-what-and-why.md:148]:
```suggestion
When using a queueing system and at the same time storing data, sometimes we need to guarantee consistency. We don't want to store data and fail in dispatching messages, nor do we wish to deliver messages and fail in storing data. We relied on distributed transactions for a long time to achieve the desired goal. However, most queuing infrastructures and cloud-enabled storages don't support distributed transactions. It takes a lot of work to implement, especially if there is a need to support multiple storage options and queuing systems. That's one of the reasons to rely upon a dedicated toolkit. For more information, please refer to the following articles:
```

File: _posts/outbox-what-and-why.md (hunk 1 above)
Comment (mauroservienti) [reply _posts/outbox-what-and-why.md:27]:
```suggestion
All this means that the above diagram only works in a cloud environment if the database server runs on the same machine where the MSMQ is installed. This sounds illogical because that means the only option to use distributed transactions is to have all participants on the same box, which is precisely what we don't want in the cloud.
```

File: _posts/outbox-what-and-why.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+3. Commit the local transaction making sure that:
+  * Data are safely stored.
+  * Messages we want to send are in the outbox storage.
+  * The incoming message id is in the database too.
+4. Send messages we want to send and mark them as sent (outside the original local transaction).
+5. Acknowledge the incoming message, deleting it from the queue.
+
+If anything between 1 and 3 fails, the local transaction gets rolled back, no changes are applied, and the endpoint will retry the incoming message.
+
+If what fails are either steps 4 or 5, the endpoint picks up the message again. The outbox already contains the message identifier, stored at the first step and committed at step three. The outbox treats it as an "already processed message for which something went wrong." In this mode, the endpoint skips processing the message and jumps to step 4, sending outgoing messages, marking them as sent, and finally acknowledging the incoming message.
```
Comment (danielmarbach) [inline _posts/outbox-what-and-why.md:86]: >  the endpoint skips processing the message

above you say 

> Handle the message from the business logic perspective

maybe it would be better to rephrase this to skip executing the business logic or something similar

Comment (mauroservienti) [reply _posts/outbox-what-and-why.md:86]:
```suggestion
If what fails are either steps 4 or 5, the endpoint picks up the message again. The outbox already contains the message identifier, stored at the first step and committed at step three. The outbox treats it as an "already processed message for which something went wrong." In this mode, the endpoint skips invoking the users' business logic and jumps to step 4, sending outgoing messages, marking them as sent, and finally acknowledging the incoming message.
```

Thanks, Daniel.
